lib/ur/usage_report.go

```diff
@@ -453,3 +474,62 @@ func cpuBenchOnce(duration time.Duration, useWeakHash bool, bs []byte) float64 {
 	d := time.Since(t0)
 	return float64(int(float64(b)/d.Seconds()/(1<<20)*100)) / 100
 }
+
+type ModelIntf interface {
+	GlobalDirectoryTree(folder, prefix string, levels int, dirsonly bool) map[string]interface{}
+	Completion(device protocol.DeviceID, folder string) model.FolderCompletion
+	Override(folder string)
+	Revert(folder string)
+	NeedFolderFiles(folder string, page, perpage int) ([]db.FileInfoTruncated, []db.FileInfoTruncated, []db.FileInfoTruncated)
+	RemoteNeedFolderFiles(device protocol.DeviceID, folder string, page, perpage int) ([]db.FileInfoTruncated, error)
+	LocalChangedFiles(folder string, page, perpage int) []db.FileInfoTruncated
+	NeedSize(folder string) db.Counts
+	ConnectionStats() map[string]interface{}
+	DeviceStatistics() map[string]stats.DeviceStatistics
+	FolderStatistics() map[string]stats.FolderStatistics
+	CurrentFolderFile(folder string, file string) (protocol.FileInfo, bool)
+	CurrentGlobalFile(folder string, file string) (protocol.FileInfo, bool)
+	ResetFolder(folder string)
+	Availability(folder string, file protocol.FileInfo, block protocol.BlockInfo) []model.Availability
+	GetIgnores(folder string) ([]string, []string, error)
+	GetFolderVersions(folder string) (map[string][]versioner.FileVersion, error)
+	RestoreFolderVersions(folder string, versions map[string]time.Time) (map[string]string, error)
+	SetIgnores(folder string, content []string) error
+	DelayScan(folder string, next time.Duration)
+	ScanFolder(folder string) error
+	ScanFolders() map[string]error
+	ScanFolderSubdirs(folder string, subs []string) error
+	BringToFront(folder, file string)
+	Connection(deviceID protocol.DeviceID) (connections.Connection, bool)
+	GlobalSize(folder string) db.Counts
+	LocalSize(folder string) db.Counts
+	ReceiveOnlyChangedSize(folder string) db.Counts
+	CurrentSequence(folder string) (int64, bool)
+	RemoteSequence(folder string) (int64, bool)
+	State(folder string) (string, time.Time, error)
+	UsageReportingStats(version int, preview bool) map[string]interface{}
+	FolderErrors(folder string) ([]model.FileError, error)
+	WatchError(folder string) error
+}
```

Comment:
A less hackish alternative is to make model, wrapper, connections, ... interfaces. E.g. instead of using `*config.Wrapper` all over, use `config.Wrapper` which is an interface and in production is "backed by `*config.wrapper`. That however means those interfaces need to be complete and thus the mocked objects need some more methods.

Reply:
Currently there are `modelIntf`, `configIntf` and `connectionsIntf` in gui.go. I see three options to handle those with split packages:

1. Create them as `ModelIntf`, ... in their respective packages (model, ...).

2. In their respective packages replace the currently exported struct used as pointers in other packages (e.g. `*config.Wrapper`) with interfaces and internalize the struct.

3. The current "solution" in this PR: Move all these interfaces to a package that is imported, but doesn't import other consumers of these interfaces.

4. Move all interfaces to a dedicated `lib/interfaces` package.

1. and 4. are very similar and in my opinion defining the interfaces in their "package of origin" is much more sensible. I still think 2. is the "correct and clean" solution, but it's too much work. 3. is simple but ugly. I'd like to go ahead with 1., but would be fine with 2. as well and as always I am open to be convinced of anything else as well.

@AudriusButkevicius ?

Reply:
I don't think we can define things in their location of origin as I though the purpose of the interfaces was to solve circular deps. If that is not the case then we can just hide the implementation by lowercasing it and always use the interface which I would be ok with. Yet as I said I was under the impression we can't do that.

Reply:
We can, these interface are purely for testing. They just allow starting services with mocked models, configs, ... instead of the real thing.

>  If that is not the case then we can just hide the implementation by lowercasing it and always use the interface which I would be ok with

Oh no, I was hoping you wouldn't like that approach xD

Reply:
Perhaps worthy a separate PR